### ISSUE TITLE
Chat - Markdown entered in email responses don’t render in the app correctly

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#4899]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Send a message from userA to userB
2. Wait for the email containing the message to be received on email by userB
3. Respond to the email and include markdown in your response

## Expected Result:
The markdown entered in the email response should render in the New Expensify app

## Actual Result:
The message via email appears in New Expensify without markdown rendered correctly

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:**
1.0.88-2
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
![Bug5212462_markdown](https://user-images.githubusercontent.com/43996225/131225796-26b223ab-97bc-4b7c-a7f5-564285029441.jpg)

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @flodnv (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by isagoico** [comment]
I also tested this using the markdowns in Gmail and message displayed only the italic markdown. 

- New Expensify: 

![image](https://user-images.githubusercontent.com/44479856/131398452-8a4d1507-184a-423e-858f-26ed62aca866.png)

- Email: 

![image](https://user-images.githubusercontent.com/44479856/131398546-30497c8e-e510-4f2d-b460-a1c00393a0fa.png)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by flodnv** [comment]
Nice shot @MelvinBot  😄 

@trjExpensify I don't think we really want to fix this? Or if we do, I'd say low priority at best?

**Comment by trjExpensify** [comment]
We can mark this issue as `external` to be worked on by a contributor.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by thienlnam** [comment]
I think this may actually end up being an internal issue since the way we handle email responses is through mailgun and this might be a case of us not parsing or converting the markdown into html

**Comment by trjExpensify** [comment]
Cool, if that transpires to be the case I can pull the job post down.

**Comment by trjExpensify** [comment]
No takers on Upwork yet, Melv.

**Comment by pROFESOR11** [comment]
### Proposal

I've been working on this issue for a while, and I wanted to share what I've got so far. If my proposal is accepted, I'd be happy to submit a PR.

As a starting point, I had to distinguish messages coming from the email replies and the application itself.
I've investigated incoming `html` values from email replies:
- If user replies via mobile app, then `html` ends with `</div><div><br /></div>`
- If user replies via browser, then `html` ends with `</div><br />`

_Note: I think it would be safer to recognize email replies via a db field or so. Anyways I tried to find a solution with what I have._

https://github.com/Expensify/App/blob/6cd518e81fb60415ba8d4d4a659fdc5dfa95b6fd/src/components/RenderHTML.js#L10-L18

```diff
+ import ExpensiMark from 'expensify-common/lib/ExpensiMark';
+ import Str from 'expensify-common/lib/str';

const RenderHTML = ({html}) => {
    const {width} = useWindowDimensions();

+   let sourceHtml = html;

+    if (html.endsWith('</div><div><br /></div>') /* mobile */
+       || html.endsWith('</div><br />') /* desktop */) {
+       const parser = new ExpensiMark();
+       sourceHtml = parser.replace(Str.stripHTML(html.replace(/((<br[^>]*>)+)/gi, '\n')));
+   }

    return (
        <RenderHTMLSource
            contentWidth={width * 0.8}
-           source={{html}}
+           source={{html: sourceHtml}}
        />
    );
};

```


Here in `RenderHTML` component, if `html` meets one of the above conditions, then first I stripped HTML and then used parser. In order to keep new lines, I replaced `<br>` tags with `\n`.

Here are the screenshots of app and web, before and after above diff applied.

App - Now             |  App - After Change
:-------------------------:|:-------------------------:
<img alt="app_now" src="https://user-images.githubusercontent.com/26159043/132227406-d0a67b79-79b2-4e79-8721-de121243f5a0.png" width="100%" /> | <img alt="app_after_change" src="https://user-images.githubusercontent.com/26159043/132227657-fdcd1472-e4f0-4809-9045-6733d593f6fc.png" width="100%" />

Web - Now             |  Web - After Change
:-------------------------:|:-------------------------:
<img alt="web_now" src="https://user-images.githubusercontent.com/26159043/132228367-eeeae2ca-fc12-4b51-bb25-f744e4b4fa73.png" width="100%" /> | <img alt="web_after_change" src="https://user-images.githubusercontent.com/26159043/132228176-6fd76a09-00f7-439b-9324-48b3ad06c978.png" width="100%" />

**Comment by MelvinBot** [comment]
@trjExpensify, @roryabraham Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by trjExpensify** [comment]
Thanks, @pROFESOR11! @roryabraham is out for Labor Day in the US, so we'll come back to you with feedback after the vacation likely tomorrow or the next day.

**Comment by roryabraham** [comment]
@pROFESOR11 Sorry for the delay in processing this proposal. Given the limited front-end resources available to you, I think you've done a good job with your proposal. But unfortunately I think we need to step back and create a more robust/holistic solution involving some back-end changes, as you and @thienlnam suggested.

Unfortunately I don't have time to flesh out a detailed implementation plan right now, but I think we can wrap chat responses sent via email in a `<p>` tag with a custom data attribute, then recognize that data attribute and process the email responses as needed in the front-end.

**Comment by roryabraham** [comment]
@trjExpensify Please take down the Upwork job for this issue.

**Comment by chiragsalian** [comment]
Started working on this today. i was playing around with `Parsedown` to see if it would help us in this scenario. It has some limitations and as I continued working on it there is a chance we don't even need it. So basically i'm trying to understand our usecase, from the issue description it talks about markdowns i.e., special characters like `**` or `~~` to determine the formatting. But from the very next example [here](https://github.com/Expensify/App/issues/4899#issuecomment-908653109) its not markdown but those are just regular formatting options. Those are most likely just plain HTML that we're either stripping out when we shouldn't be or updating App to those elements.

Eg of markdown bold: `*Test Me*`
Eg of formatted style for bold: `<b>Test Me</b>`

If you bold a text in gmail we'll receive its formatted style. Our app currently does not recognize `<b>` and instead uses `<strong>` for bold.

So anyway I think we should not be looking at converting markdown from the emails and instead just allowing formatted email styles to show. Since the latter is more common and I'm not sure why would anyone send an email with markdown text. Let me know if that makes sense or if you guys feel differently.

**Comment by thienlnam** [comment]
My bad, my comment about parsing emails into markdown was not accurate

> instead just allowing formatted email styles to show

We already do this mostly, we get HTML from emails but have to strip them of quoted content and images (since we process those separately and not inline) [here](https://github.com/Expensify/Web-Expensify/blob/b8233a0d4c05d6efcb1d58928f04532ba472cd88/lib/EmailUtils.php#L132-L132).

Quoted content being stripped is expected for previous email replies but we may need to update our parser to only strip out previous email content instead of all quoted content

> Our app currently does not recognize <b> and instead uses <strong> for bold.
This is probably something we can add into expensimark

**Comment by thienlnam** [comment]
I was chatting with @chiragsalian about this 1:1 but I don't think we should support markdown entered in emails like the issue description mentioned.  Instead, they should just use the included formatting that email providers allow.  I think this Problem/Solution to this would be:

Problem: Formatting used in emails is not being formatted correctly in the app
Solution: Add support for <b> and <strike> tags in Expensimark

**Comment by chiragsalian** [comment]
Got a bit sidetracked due to other priorities and then I've been OOO this week cause I'm helping with the career fair. I should get back to this issue soon.

**Comment by chiragsalian** [comment]
Demoting to monthly since i'm focusing on other tasks while this still has n6-hold on it.

**Comment by chiragsalian** [comment]
Was focusing on other priorities. Now that my plate is clear i should get back to this soon.

**Comment by chiragsalian** [comment]
Fixed, I'm adding support for bold, italics, and strikethrough from emails. The changes were easier than I thought, I just had to update our react-native-render-html since support for bold was added recently. Italics always worked. Strikethrough was stripped in the backend so I'm allowing it now and react-native-render-html works perfectly with strike tags as well.

If there are additional styles needed we can add them in on a case-to-case basis.

PR in review.
